File: src/test/java/com/example/usercrud/Business/Service/UserServiceTest.java
package com.example.usercrud.Business.Service;

import com.example.usercrud.Business.Entity.User;
import com.example.usercrud.Persistance.Repository.UserRepository;
import org.aspectj.lang.annotation.Before;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.mockito.InjectMocks;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.test.annotation.DirtiesContext;
import org.springframework.test.context.TestPropertySource;

import java.util.Optional;

import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
import static org.junit.jupiter.api.Assertions.*;

@SpringBootTest
@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
class UserServiceTest {

    @Autowired
    private UserRepository userRepoTest;

    @Autowired
    private UserService userServiceTest;


    @Test
    @DisplayName("Delete not existing user by id")
    void itShouldDeleteNotExistingUserById() {
        //given
        Long id = -1L;

        //when
        Optional<User> expectedEmpty = userServiceTest.deleteById(id);

        //then
        assertThat(expectedEmpty).isEmpty();
    }
    @Test
    @DisplayName("Delete existing user by id")
    void itShouldDeleteExistingUserById(){
        //given
        User user = new User(1L, "firstname", "secondname", "middlename", "country", (short) 0, "email@email.email");
        userRepoTest.save(user);

        //when
        Optional<User> exceptedPresent = userServiceTest.deleteById(1L);

        //then
        assertThat(exceptedPresent).isPresent();
    }

    @Test
    @DisplayName("Delete not existing user by email")
    void itShouldDeleteNotExistingUserByEmail() {
        //given
        String notExistingEmail = "notexisting@email.email";

        //when
        Optional<User> expectedEmpty = userServiceTest.deleteByEmail(notExistingEmail);

        //then
        assertThat(expectedEmpty).isEmpty();
    }


    @Test
    @DisplayName("Delete existing user by email")
    void itShouldDeleteExistingUserByEmail() {
        //given
        User user = new User(1L, "firstname", "secondname", "middlename", "country", (short) 0, "email@email.email");
        userRepoTest.save(user);

        //when
        Optional<User> exceptedPresent = userServiceTest.deleteByEmail("email@email.email");

        //then
        assertThat(exceptedPresent).isPresent();
    }

    @Test
    @DisplayName("Check if not existing user exists by email")
    void itShouldCheckIfNotExistingUserExistsByEmail() {
        //given
        String notExistingEmail = "notexisting@email.email";

        //when
        boolean expectedFalse = userServiceTest.existsByEmail(notExistingEmail);

        //then
        assertThat(expectedFalse).isFalse();
    }

    @Test
    @DisplayName("Check if existing user exists by email")
    void itShouldCheckIfExistingUserExistsByEmail() {
        //given
        String existingEmail = "email@email.email";
        User user = new User(1L, "firstname", "secondname", "middlename", "country", (short) 0, existingEmail);
        userServiceTest.addUser(user);

        //when
        boolean exceptedTrue = userServiceTest.existsByEmail(existingEmail);

        //then
        assertThat(exceptedTrue).isTrue();
    }

    @Test
    @DisplayName("Update not existing user by email")
    void itShouldUpdateNotExistingUserByEmail() {
        //given
        String notExistingEmail = "notexisting@email.email";
        User notExistingUser = new User(1L, "firstname", "secondname", "middlename", "country", (short) 0, notExistingEmail);

        //when
        Optional<User> exceptedEmpty = userServiceTest.updateByEmail(notExistingEmail, notExistingUser);

        //then
        assertThat(exceptedEmpty).isEmpty();
    }

    @Test
    @DisplayName("Update existing user by email")
    void itShouldUpdateExistingUserByEmail() {
        //given
        String existingEmail = "existing@email.email";
        User existingUser = new User(1L, "firstname", "secondname", "middlename", "country", (short) 0, existingEmail);
        userServiceTest.addUser(existingUser);
        existingUser.setFirstName("newFirstName");
        existingUser.setLastName("newLastName");
        existingUser.setMiddleName("newMiddleName");
        existingUser.setEmail("newEmail");
        existingUser.setGender((short)1);

        //when
        Optional<User> exceptedPresent = userServiceTest.updateByEmail(existingEmail, existingUser);

        //then
        assertThat(exceptedPresent).isPresent();
        assertThat(exceptedPresent.get().getFirstName()).isEqualTo(existingUser.getFirstName());
        assertThat(exceptedPresent.get().getLastName()).isEqualTo(existingUser.getLastName());
        assertThat(exceptedPresent.get().getEmail()).isEqualTo(existingUser.getEmail());
        assertThat(exceptedPresent.get().getMiddleName()).isEqualTo(existingUser.getMiddleName());
        assertThat(exceptedPresent.get().getGender()).isEqualTo(existingUser.getGender());
    }

    @Test
    @DisplayName("Update not existing user by id")
    void itShouldUpdateNotExistingUserById() {
        Long notExistingId = 1L;
        User notExistingUser = new User(1L, "firstname", "secondname", "middlename", "country", (short) 0, "notexist@email.email");

        //when
        Optional<User> exceptedEmpty = userServiceTest.updateById(notExistingId, notExistingUser);

        //then
        assertThat(exceptedEmpty).isEmpty();
    }

    @Test
    @DisplayName("Update existing user by id")
    void itShouldUpdateExistingUserById() {
        //given
        Long existingId = 1L;
        User existingUser = new User(1L, "firstname", "secondname", "middlename", "country", (short) 0, "existingEmail@email.email");
        userServiceTest.addUser(existingUser);
        existingUser.setFirstName("newFirstName");
        existingUser.setLastName("newLastName");
        existingUser.setMiddleName("newMiddleName");
        existingUser.setEmail("newEmail");
        existingUser.setGender((short)1);

        //when
        Optional<User> exceptedPresent = userServiceTest.updateById(existingId, existingUser);

        //then
        assertThat(exceptedPresent).isPresent();
        assertThat(exceptedPresent.get().getFirstName()).isEqualTo(existingUser.getFirstName());
        assertThat(exceptedPresent.get().getLastName()).isEqualTo(existingUser.getLastName());
        assertThat(exceptedPresent.get().getEmail()).isEqualTo(existingUser.getEmail());
        assertThat(exceptedPresent.get().getMiddleName()).isEqualTo(existingUser.getMiddleName());
        assertThat(exceptedPresent.get().getGender()).isEqualTo(existingUser.getGender());
    }
}

File: src/test/kotlin/cloud/folium/usercrud/unit/UserServiceTest.kt
package cloud.folium.usercrud.unit

import cloud.folium.usercrud.business.entity.User
import cloud.folium.usercrud.business.service.UserService
import cloud.folium.usercrud.persistance.repository.UserRepositoryHibernate
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.extensions.spring.SpringExtension
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import io.mockk.*
import io.mockk.impl.annotations.MockK
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest

@SpringBootTest
class UserServiceTest : BehaviorSpec() {
    override fun extensions() = listOf(SpringExtension)

    @Autowired
    lateinit var userService: UserService

    val testUser = User(1L, "firstName", "lastName", "middleName", "country", 1, "email@email.email", null)
    val testNewUser = User(1L, "newFirstName", "newLastName", "newMiddleName", "newCountry", 2, "newEmail@email.email", null)
    @MockK
    val userRepository = mockk<UserRepositoryHibernate>()

    init {
        /*
        given ("Repository returns false for existsById()") {
            every {
                userRepository.existsById(any())
            } returns false
            userService.setRepository(userRepository)
            `when` ("calling service to delete not existing user by ID") {
                then ("should return false") {
                    userService.deleteById(1L) shouldBe false
                }
            }
        }

        given("Repository returns true for existsById()") {
            every {
                userRepository.existsById(any())
            } returns true
            userService.setRepository(userRepository)
            `when` ("calling service to delete existing user by ID") {
                then ("should return true") {
                    userService.deleteById(1L) shouldBe true
                }
            }
        }

         */

        given ("Repository returns false for existsByEmail()") {
            every {
                userRepository.existsByEmail(any())
            } returns false
            userService.setRepository(userRepository)
            `when` ("calling service to delete user by Email") {
                then ("should return false") {
                    userService.deleteByEmail("email@email.email") shouldBe false
                }
            }
        }

        given("Repository returns true for existsByEmail()") {
            every {
                userRepository.existsByEmail(any())
            } returns true
            every {
                userRepository.deleteByEmail(any())
            } just Runs
            userService.setRepository(userRepository)
            `when` ("calling service to delete user by Email") {
                then ("should return true") {
                    userService.deleteByEmail("email@email.email") shouldBe true
                }
            }
        }

        given("Repository returns true for existsByEmail()") {
            every {
                userRepository.existsByEmail(any())
            } returns true
            userService.setRepository(userRepository)
            `when` ("calling service to check if user exists") {
                then ("should return true") {
                    userService.existsByEmail("email@email.email") shouldBe true
                }
            }
        }

        given("Repository returns false for existsByEmail()") {
            every {
                userRepository.existsByEmail(any())
            } returns false
            userService.setRepository(userRepository)
            `when` ("calling service to check if user exists") {
                then ("should return false") {
                    userService.existsByEmail("email@email.email") shouldBe false
                }
            }
        }

        given("Repository returns null for findByEmail()") {
            every {
                userRepository.findByEmail(any())
            } returns null
            userService.setRepository(userRepository)
            `when` ("update user") {
                then ("should return null") {
                    userService.updateByEmail("email@email.email", testUser) shouldBe null
                }
            }
        }

        given("Repository returns testUser for findByEmail()") {
            every {
                userRepository.findByEmail(any())
            } returns testUser
            userService.setRepository(userRepository)
            `when` ("update user") {
                val updatedUser = userService.updateByEmail("email@email.email", testNewUser)
                then ("received object's fields should match with sent user object fields") {
                    updatedUser shouldNotBe null
                    updatedUser?.apply {
                        firstName shouldBe "newFirstName"
                        lastName shouldBe "newLastName"
                        middleName shouldBe "newMiddleName"
                        email shouldBe "newEmail@email.email"
                        country shouldBe "newCountry"
                        gender shouldBe 2
                    }
                }
            }
        }

        given("Repository return null for findByEmail()") {
            every {
                userRepository.findByEmail(any())
            } returns null
            `when` ("update user") {
                val updatedUser = userService.updateByEmail("email@email.email", testNewUser)
                then("received object should be null") {
                    updatedUser shouldBe null
                }
            }
        }
/*
        given("Repository returns testUser for findById()") {
            every {
                userRepository.findById(any())
            } returns testUser
            userService.setRepository(userRepository)
            `when` ("update user") {
                val updatedUser = userService.updateById(1L, testNewUser)
                then ("received object's fields should match with sent user object fields") {
                    updatedUser shouldNotBe null
                    updatedUser?.apply {
                        firstName shouldBe "newFirstName"
                        lastName shouldBe "newLastName"
                        middleName shouldBe "newMiddleName"
                        email shouldBe "newEmail@email.email"
                        country shouldBe "newCountry"
                        gender shouldBe 2
                    }
                }
            }
        }

        given("Repository return null for findById()") {
            every {
                userRepository.findById(any())
            } returns null
            `when` ("update user") {
                val updatedUser = userService.updateById(1L, testNewUser)
                then("received object should be null") {
                    updatedUser shouldBe null
                }
            }
        }

 */
    }

}

